### PR TITLE
Feature 1695 ensemble single file (gen_ens_prod)

### LINKED ISSUE #1695: Enhance Ensemble-Stat and Gen-Ens-Prod to read all ensemble members from a single input file
## Describe the New Feature ##
In our discussions with CPC, we've been presented with NMME netCDF data files where all of the ensemble members are located in the same data structure in 1 file. CPC confirmed this is the normal structure for ensemble files, and has been encountered during work with other entities (e.g. NRL). While Python embedding is currently a viable solution to this file design type, there's a desire for MET's EnsembleStat tool to do this natively. 

### Acceptance Testing ###
Sample NMME file can be found on kiowa, under /d1/projects/MET/MET_test_data/CPC_NMME. Variables of interest are `ensmem(ensmem)` and `fcst(ensmem,target,lat,lon)`
Tests should successfully read in all ensemble members in datafile into EnsembleStat with available processing options mirroring what was available for individual member files.

### Time Estimate ###
*Estimate the amount of work required here.*
Unknown until later date, input from Engineers

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
None

### Funding Source ###
TBD when on a beta schedule

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [x] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Review **projects** and select relevant **Repository** and **Organization** ones or add "alert:NEED PROJECT ASSIGNMENT" label
- [x] Select **milestone** to next major version milestone or "Future Versions"

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## New Feature Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, **Milestone**, and **Linked issues**
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by georgemccabe**
Added support for running gen_ens_prod on a single file that contains all ensemble members

Changes include:
* Modified string_value function to substitute values for MET_ENS_MEMBER_ID and return ConcatString instead of ConcatString*
* Created class EnsVarInfo to store all info to read ensemble files/fields
* Refactored logic in gen_ens_prod to read from EnsVarInfo

Note: EnsVarInfo class will be moved into the library code in the next pull request so that it can also be used by ensemble_stat

## Expected Differences ##

- [X] Do these changes introduce new tools, command line arguments, or configuration file options? **[Yes]**</br>

New configuration file options for gen_ens_prod: ens_member_ids and control_id

- [X] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>

## Pull Request Testing ##

- [X] Describe testing already performed for these changes:</br>

Added unit tests that demonstrate new functionality

- [X] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

Review code changes, run gen_ens_prod unit tests and ensure that they all pass (new data for testing is available on seneca and kiowa)

- [X] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [X] Do these changes include sufficient testing updates? **[Yes]**

- [X] Will this PR result in changes to the test suite? **[Yes]**</br>
New output will be generated from 4 new unit tests. No other differences should result.

- [X] Please complete this pull request review by **1/11/2022**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [X] Complete the PR definition above.
- [X] Ensure the PR title matches the feature or bugfix branch name.
- [X] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Linked issues** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by georgemccabe** [comment]
Based on discussion in a NOAA telecon, I updated the error message if the valid thresh is not met so that it is more clear what values are being compared when the threshold is not met.

Old error message:
> ERROR  : process_ensemble() -> 2 of 3 missing fields for "fcst(1,0,\*,\*)" exceeds the maximum allowable specified by "ens.ens_thresh" (1) in the configuration file.

New error message:
> ERROR  : process_ensemble() -> 1 of 3 (0.333333) fields found for "fcst(1,0,\*,\*)" does not meet the threshold specified by "ens.ens_thresh" (1) in the configuration file.

File: met/data/config/GenEnsProdConfig_default [hunk 1]
```diff
@@ -57,6 +57,13 @@ ens = {
    ];
 }
 
+//
+// IDs for ensemble members
+// Only set if processing a single file with all ensembles
```
Comment (JohnHalleyGotway) [inline met/data/config/GenEnsProdConfig_default:62]:
```suggestion
// Only set if processing a single file with all ensemble members
```

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of these changes (BUT PLEASE SEE NOTE BELOW)

I have one minor recommendation. In order to avoid lots of versions of config file contents floating around... whenever I add a new config option to one of the config file, I also add it (with its default value) to all instances of that config file in the "test/config" directory. This isn't a "rule", just a recommendation. So I'd recommend adding those 2 new config options to this existing file:

https://github.com/dtcenter/MET/blob/feature_1695_ensemble_single_file/test/config/GenEnsProdConfig

I ran a full regression test vs develop for this feature branch in kiowa:/d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta5/feature_1695_gen_ens_prod.

All existing output is the same, but it introduces 4 new output files:
```
ERROR: folder MET-develop/test_output missing 4 files
    gen_ens_prod/gen_ens_prod_SINGLE_FILE_GRIB_NO_CTRL.nc 
    gen_ens_prod/gen_ens_prod_SINGLE_FILE_GRIB_WITH_CTRL.nc 
    gen_ens_prod/gen_ens_prod_SINGLE_FILE_NC_NO_CTRL.nc 
    gen_ens_prod/gen_ens_prod_SINGLE_FILE_NC_WITH_CTRL.nc 
```

Checking these new outputs, I see:
1. The output fields look reasonable via ncview.
2. Diffing them with ncdiff produces the expected result:
```
/usr/local/nco/bin/ncdiff gen_ens_prod/gen_ens_prod_SINGLE_FILE_GRIB_NO_CTRL.nc gen_ens_prod/gen_ens_prod_SINGLE_FILE_GRIB_WITH_CTRL.nc grib_diff.nc
```

The standard deviation fields have 0 differences, but the rest do. The mean/mean +/- stadev fields have lot of diffs because the control (when specified) is included in the computation of the mean. The min/max/range fields have fewer diffs... only at grid points where the control has the extreme value.
3. I noted the same result in the ncdiff output for the NetCDF NO/WITH_CTRL output files.

Reviewing the code changes, I note that we:
- Add to gen_ens_prod new config options named ens_memeber_ids and control_id.
- Added user's guide documentation for these including examples.
- Changes to the vx_config library to parse them.
- Change to the gen_ens_prod to no longer interpret the number of input files as the number of ensemble members to be processed.
- Change e->string_value() to return a ConcatString instance instead of a pointer to one. That's need to be able to substitute in the ensemble member id string value.

**Comment by georgemccabe** [comment]
I updated the other test config file for GenEnsProd to include the new default values as suggested in the PR review.

**Pull request merged**